### PR TITLE
[REVIEW] Execution policy class

### PR DESCRIPTION
**Pull request by viclafargue**
Changes currently planned for the RAFT code include updating the RAFT handle for it to store a singleton of a thrust execution policy. An execution policy class would prove useful to write a clearer code. This PR is a proof of concept for it.

cc @divyegala

### CODE REVIEW

**Comment by jrhemstad** [comment]
> to store a singleton of a thrust execution policy

I'm curious why you wish to do this. Are you concerned about the overhead of constructing an execution policy on the fly with the existing `rmm::exec_policy` function?

**Comment by viclafargue** [comment]
> > to store a singleton of a thrust execution policy
> 
> I'm curious why you wish to do this. Are you concerned about the overhead of constructing an execution policy on the fly with the existing `rmm::exec_policy` function?

We are planning on adding a `raft::handle_t::get_thrust_policy()` method to the RAFT handle and use it everywhere a thrust execution policy is required. This would be to provide an abstraction over the `rmm::exec_policy` and would prevent any depreciation. Indeed, if the RMM code is modified, we would only have to update this method in the RAFT handle. But this idea is still in discussion. The idea of having a singleton is for both efficiency and also to be coherent in the way resources are created and provided to developers in the RAFT handler. Indeed, cuBlas, cuSparse and cuSolver handles are stored as singleton as well.

**Comment by jrhemstad** [comment]
`raft::handle_t::get_thrust_policy()` could just return `rmm::exec_policy` as it is today, right?

I don't think you have to worry about the overhead of constructing the execution policy as it will all get inlined/optimized away anyways.

That said, I'm not opposed to an `exec_policy` class, but I don't want to have to change all of the existing calls to `rmm::exec_policy` to `rmm::create_exec_policy`.

**Comment by divyegala** [comment]
@jrhemstad if the overhead isn't very high of creating, then I agree that `raft::handle_t::get_thrust_policy()` could just call `rmm::exec_policy()`. But it would be nice to hide that behind an `rmm::exec_policy` class that will handle deprecations to both the RMM and thrust APIs. Furthermore, I agree that a factory method here isn't the right approach.

In that case, `rmm::exec_policy` could be a class that has an implicit conversion operator to thrust's policy class, and is coupled with a stream view instance. Would that work?

**Comment by viclafargue** [comment]
I renamed the `rmm::exec_policy` function back to its original name, and implemented the suggested class.

**Comment by harrism** [comment]
Why is deprecation such a concern? `rmm::exec_policy` has only had one deprecation in the life of the library, and I think the current API should be quite stable.

**Comment by harrism** [comment]
To pass the style checks please merge the latest from upstream into your PR branch.

**Comment by viclafargue** [comment]
> Why is deprecation such a concern? `rmm::exec_policy` has only had one deprecation in the life of the library, and I think the current API should be quite stable.

It might indeed never come useful if the API is left unchanged. However, I guess it's probably better to be prepared for such case. The change might come from an update in Thrust independently from our will.

File: include/rmm/exec_policy.hpp [hunk 1]
```diff
@@ -38,4 +42,13 @@ inline auto exec_policy(cuda_stream_view stream             = cuda_stream_defaul
   return thrust::cuda::par(rmm::mr::thrust_allocator<char>(stream, mr)).on(stream.value());
 }
 
+struct exec_policy_t {
```
Comment (divyegala) [inline include/rmm/exec_policy.hpp:45]: Can't this class just be named `exec_policy`?

File: include/rmm/exec_policy.hpp [hunk 2]
```diff
@@ -38,4 +42,13 @@ inline auto exec_policy(cuda_stream_view stream             = cuda_stream_defaul
...
+  explicit exec_policy_t(cuda_stream_view stream) : stream_(stream) {}
+
+  operator thrust_exec_policy_t() { return exec_policy(stream_); }
```
Comment (divyegala) [inline include/rmm/exec_policy.hpp:48]: With the above change, you could return here `return thrust::cuda::par(rmm::mr::thrust_allocator<char>(stream, mr)).on(stream.value());`

**Review by divyegala — COMMENTED** [review]
I believe with this, we can have a non-breaking API update without changing the way this was originally working. cc @jrhemstad @harrism who will be able to help us best here

**Comment by harrism** [comment]
It would be nice to follow Divye's suggestion to make this a replacement for the existing function that works identically. Also, I believe `_t` names are meant to be reserved for use by standard library types.

**Comment by viclafargue** [comment]
I removed the function and renamed the class as `exec_policy`. After some discussion with Divye, it appeared that the conversion function couldn't work because of templates in thrust. Instead the class is now designed to inherit from the thrust execution policy.

**Review by divyegala — APPROVED** [review]
This looks good for RAFT and doesn't break the existing code. LGTM!

**Comment by viclafargue** [comment]
rerun test

**Comment by viclafargue** [comment]
rerun tests

**Comment by viclafargue** [comment]
The PR should be ready to be merged.

**Review by harrism — APPROVED** [review]
Thanks!

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**